### PR TITLE
Refactor model objects

### LINKED ISSUE #3: Refactor model objects
Explore integrating the [Immutables](http://immutables.github.io/) library to replace the boilerplate Builder code throughout the models in the codebase.

### PR DESCRIPTION
**Pull request by macklinu**
This commit performs some major refactoring of the app’s models. A lot
of repeated code was removed by code generation tools AutoValue and
AutoValue-Parcel.

https://github.com/google/auto/tree/master/value
https://github.com/rharter/auto-value-parcel

Now, models are abstract classes that basically define the models’
public APIs. Most models also have an inner abstract Builder class.
Both are generated by AutoValue.

Model classes that need custom parceling behavior are annotated with
`@ParcelAdapter` and a custom TypeAdapter parceling implementation.
This allows AutoValue-Parcel to properly generate parceling code to
maintain the Parcelable interface of our models.

To accommodate the updated model interfaces, Gson deserializers have
been updated and/or created to properly deserialize JSON from the API
into the proper models.

Resolves #3

### CODE REVIEW

**Comment by macklinu** [comment]
@crebma I'm assigning this to you. I know this is a large PR, but if you wouldn't mind either doing a quick QA or asking questions or just saying "do it up", that'd be cool, just so I'm not introducing a bunch of changes on the sly.

**Comment by crebma** [comment]
I will check it out tonight!

File: build.gradle [hunk 1]
```diff
@@ -28,7 +30,7 @@ allprojects {
             otto                : "1.3.8",
             gson                : "2.3.1",
             jodaTime            : "2.8.2",
-            guava               : "18.0",
+            guava               : "19.0",
```
Comment (crebma) [inline build.gradle]: For some reason, this is my favorite line in the whole diff.

**Comment by crebma** [comment]
This looks really awesome!!

**Pull request merged**